### PR TITLE
fix: e2e test coverage setup

### PR DESCRIPTION
**Pull request by omkarK06**
## Release Notes

- **New Features**
	- Added support for code coverage reporting with the integration of `vite-plugin-istanbul`.
	- Introduced a new configuration file for code coverage settings.

- **Improvements**
	- Updated project to use ECMAScript modules for better compatibility.
	- Enhanced UI test coverage with additional test cases and improved error handling.
	- Refactored test suites to utilize a custom module for better organization and maintainability.
	- Improved the login process and dashboard loading in test cases for better reliability.

- **Bug Fixes**
	- Resolved issues related to import statements affecting test execution.

- **Documentation**
	- Updated `.gitignore` to exclude coverage directories from version control.

These updates enhance the functionality, maintainability, and reliability of the application and its testing framework.

### CODE REVIEW

**Review by neha00290 — APPROVED** [review]

File: .github/workflows/playwright.yml [hunk 1]
```diff
@@ -144,9 +147,64 @@
...
+      - name: Sanitize Test Filename
+        shell: bash 
+        run: |
+          # Assign the matrix variable to a Bash variable
+          SANITIZED_FILENAME="${{ matrix.testfilename }}"
+          # Replace all '/' characters with '_'
+          SANITIZED_FILENAME="${SANITIZED_FILENAME//\//_}"
+          # Export as a GitHub Actions environment variable
+          echo "SANITIZED_FILENAME=$SANITIZED_FILENAME" >> $GITHUB_ENV
+
```
[the remark this answers is not in the document]
Comment (omkarK06) [reply .github/workflows/playwright.yml:159]: Works fine

File: .github/workflows/playwright.yml [hunk 2]
```diff
@@ -144,9 +147,64 @@ jobs:
...
       - uses: actions/upload-artifact@v4
         if: always()
         with:
-          name: playwright-report
-          path: playwright-report/
-          retention-days: 30
+            name: playwright-report
+            path: playwright-report/
+            retention-days: 30
+
```
[the remark this answers is not in the document]
Comment (omkarK06) [reply .github/workflows/playwright.yml:176]: Fixed

File: tests/ui-testing/playwright-tests/reports.spec.js [hunk 3]
```diff
@@ -1,4 +1,4 @@
-import { test, expect } from "@playwright/test";
+import { test, expect } from "./baseFixtures";
 import logData from "../../ui-testing/cypress/fixtures/log.json";
 import { log } from "console";
```
[the remark this answers is not in the document]
Comment (omkarK06) [reply tests/ui-testing/playwright-tests/reports.spec.js:3]: Removed

File: tests/ui-testing/playwright-tests/indexquery.spec.js [hunk 4]
```diff
@@ -1,4 +1,4 @@
-import { test, expect } from "@playwright/test";
+import { test, expect } from "./baseFixtures";
```
[the remark this answers is not in the document]
Comment (omkarK06) [reply tests/ui-testing/playwright-tests/indexquery.spec.js:1]: Fixed Import

File: tests/ui-testing/playwright-tests/baseFixtures.js [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+    );
+    await fs.promises.mkdir(istanbulCLIOutput, { recursive: true });
+    await context.exposeFunction('collectIstanbulCoverage', (coverageJSON) => {
+      if (coverageJSON)
+        fs.writeFileSync(path.join(istanbulCLIOutput, `playwright_coverage_${generateUUID()}.json`), coverageJSON);
+    });
+    await use(context);
+    for (const page of context.pages()) {
+      await page.evaluate(() => (window).collectIstanbulCoverage(JSON.stringify((window).__coverage__)))
+    }
```
[the remark this answers is not in the document]
Comment (omkarK06) [reply tests/ui-testing/playwright-tests/baseFixtures.js:27]: Fixed

File: tests/ui-testing/playwright-tests/baseFixtures.js [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+  return crypto.randomBytes(16).toString('hex');
+}
+
+export const test = baseTest.extend({
+  context: async ({ context }, use) => {
+    await context.addInitScript(() =>
+      window.addEventListener('beforeunload', () =>
+        (window).collectIstanbulCoverage(JSON.stringify((window).__coverage__))
+      ),
+    );
```
[the remark this answers is not in the document]
Comment (omkarK06) [reply tests/ui-testing/playwright-tests/baseFixtures.js:18]: Done

File: tests/ui-testing/playwright-tests/baseFixtures.js [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+    await fs.promises.mkdir(istanbulCLIOutput, { recursive: true });
+    await context.exposeFunction('collectIstanbulCoverage', (coverageJSON) => {
+      if (coverageJSON)
+        fs.writeFileSync(path.join(istanbulCLIOutput, `playwright_coverage_${generateUUID()}.json`), coverageJSON);
+    });
```
[the remark this answers is not in the document]
Comment (omkarK06) [reply tests/ui-testing/playwright-tests/baseFixtures.js:23]: Refactored

File: web/vite.config.mts [hunk 8]
```diff
@@ -108,9 +110,17 @@ export default defineConfig({
...
       sassVariables: "src/styles/quasar-variables.sass",
     }),
+    istanbul({
+      include: "src/*",
+      exclude: ["node_modules", "test/"],
+      extension: [".js", ".ts", ".vue"],
+      requireEnv: false,
+      cypress: true,
+      forceBuildInstrument: true,
+    }),
```
[the remark this answers is not in the document]
Comment (omkarK06) [reply web/vite.config.mts]: Refactored

File: web/vite.config.ts [hunk 9]
```diff
@@ -27,6 +27,8 @@ import fs from "fs-extra";
 import monacoEditorPlugin from "vite-plugin-monaco-editor";
 import visualizer from "rollup-plugin-visualizer";
 
+import istanbul from "vite-plugin-istanbul";
+
```
[the remark this answers is not in the document]
Comment (omkarK06) [reply web/vite.config.ts]: Not required

File: .github/workflows/playwright.yml [hunk 10]
```diff
@@ -144,9 +147,64 @@ jobs:
...
+      - name: Upload Coverage Data
+        uses: actions/upload-artifact@v4
+        with:
+          name: playwright-coverage-${{ env.SANITIZED_FILENAME }}
+          path: tests/ui-testing/.nyc_output
+          include-hidden-files: true
+
+      - name: Generate Coverage Report
+        run: cd tests/ui-testing && npx nyc report
+
```
[the remark this answers is not in the document]
Comment (omkarK06) [reply .github/workflows/playwright.yml:169]: This is fine

**Pull request merged**